File: test/today-command-tests.js
'use strict';

const assert = require('chai').assert;
const moment = require('moment');

const TodayCommand = require('../lib/commands/today-command');
const Menu = require('../data/models/menu');
const MongoConnect = require('../data/mongo-connect');

describe('TodayCommand', () => {

  before(() => {
    MongoConnect.initialize();
  });

  afterEach(async () => {
    await Menu.remove({});
  });

  describe('#handle()', () => {

    it('should return today\'s menu', async () => {
      await Menu.create({ date: moment().format('YYYYMMDD'), meals: ['Foo', 'Bar'] });
      const expected = `Today is the *${moment().format('DD.MM.YYYY')}* and the meals are:\n  Foo\n  Bar\n`;
      const response = await new TodayCommand(null).handle();
      assert.equal(response, expected);
    });

    it('should return notice when there is no data', async () => {
      const response = await new TodayCommand(null).handle();
      assert.equal(response, 'I don\'t know which meals are being served today!');
    });
  });
});

File: src/Deathstar.Canteen.Tests/Units/TodayCommandTests.cs
using System;
using System.Linq.Expressions;
using System.Threading;
using System.Threading.Tasks;
using Deathstar.Canteen.Commands;
using Deathstar.Canteen.Commands.Abstractions;
using Deathstar.Canteen.Persistence;
using Deathstar.Canteen.Slack;
using Deathstar.Canteen.Tests.Mocks;
using NSubstitute;
using Xunit;

namespace Deathstar.Canteen.Tests.Units
{
	public class TodayCommandTests
	{
		private readonly IMenuCollection menuCollection = Substitute.For<IMenuCollection>();
		private readonly ISlackbot slackbot = Substitute.For<ISlackbot>();
		private readonly ICommand command;

		public TodayCommandTests() => command = new TodayCommand(menuCollection, slackbot);

		[Fact]
		public async Task TheHandleMethodShouldReturnErrorNoticeWhenNoMenuisPresentAsync()
		{
			// Arrange
			var commandMessage = new FakeCommandMessage(null, string.Empty);

			// Act
			await command.HandleAsync(commandMessage, CancellationToken.None);

			// Assert
			slackbot.Received().SendMessage(string.Empty, "I don't know which meals are being served today!");
		}

		[Fact]
		public async Task TheHandleMethodShouldReturnTodaysMenuAsync()
		{
			// Arrange
			var menu = new Menu { Date = DateTime.Today.ToString("yyyyMMdd"), Meals = new[] { "Foo", "Bar" } };
			var commandMessage = new FakeCommandMessage(null, string.Empty);

			menuCollection.SingleOrDefaultAsync(Arg.Any<Expression<Func<Menu, bool>>>(), CancellationToken.None).Returns(menu);

			// Act
			await command.HandleAsync(commandMessage, CancellationToken.None);

			// Assert
			slackbot.Received().SendMessage(
				string.Empty,
				$"Today is the *{DateTime.Today:dd.MM.yyyy}* and the meals are:{Environment.NewLine}{menu}");
		}
	}
}
